clippy_lints/src/utils/hir_utils.rs

```diff
@@ -586,4 +601,100 @@ impl<'a, 'tcx: 'a> SpanlessHash<'a, 'tcx> {
             },
         }
     }
+
+    pub fn hash_lifetime(&mut self, lifetime: &Lifetime) {
+        if let LifetimeName::Param(ref name) = lifetime.name {
+            match name {
+                ParamName::Plain(ref ident) => {
+                    ident.name.hash(&mut self.s);
+                },
+                ParamName::Fresh(ref size) => {
+                    size.hash(&mut self.s);
+                },
+                _ => {},
+            }
+        }
+    }
+
+    pub fn hash_ty(&mut self, ty: &Ty) {
+        std::mem::discriminant(&ty.node).hash(&mut self.s);
+        match ty.node {
+            Ty::Slice(ty) => {
```

Comment:
These have to be 
```suggestion
            ty::Slice(ty) => {
```

Reply:
So they still don't work when changed. Is there anywhere I can look in the source or doc.rs to find out how to get `ty::Slice` etc instead of using `TyKind::Slice` like I was before?

Reply:
Oh I think I know what's going on here now. Do you want to use `hir::Ty` and `hir::TyKind` or `ty::Ty` and `ty::TyKind`?

In the first case, you have to use the `TyKind` enum (and I'm sorry for the misleading suggestion). In the second case you need to change some of your imports, since writing `Ty` is ambiguous.

Reply:
I think hir::Ty, if that has the type information required to hash/compare the types without taking into account span information

Reply:
You can access the doc for rustc internals through https://forge.rust-lang.org

For this relevant docs:
https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/struct.Ty.html
https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/type.Ty.html

Reply:
Ok, then you have to remove the `ty::Ty` import and use `TyKind::_`. Sorry for the misleading suggestions!